### PR TITLE
feat(): add aria live announcer

### LINKED ISSUE #106: Build aria-live announcer service / component

### PR DESCRIPTION
**Pull request by DevVersion**
Fixes #106

### CODE REVIEW

File: src/components/aria-live/aria-live.scss [hunk 1]
```diff
@@ -0,0 +1,17 @@
+:host {
+
+  // This hides the announcer element visually.
+  // That means it's still accessible for screen-readers but not visible.
+  ._md-aria-announcer {
```
Comment (jelbourn) [inline src/components/aria-live/aria-live.scss]: You can omit the underscore from the name.

File: src/components/aria-live/aria-live.scss [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    border: 0;
+    clip: rect(0 0 0 0);
+    height: 1px;
+    margin: -1px;
+    overflow: hidden;
+    padding: 0;
+    position: absolute;
+    text-transform: none;
+    width: 1px;
```
Comment (jelbourn) [inline src/components/aria-live/aria-live.scss]: Can you break this styling out to a scss mixin called `md-visually-hidden` ?

File: src/components/aria-live/aria-live.html [hunk 3]
```diff
@@ -0,0 +1,2 @@
+<ng-content></ng-content>
```
Comment (jelbourn) [inline src/components/aria-live/aria-live.html]: I don't think this should actually have an `ng-content`

File: src/components/aria-live/aria-live.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  private announcerEl: HTMLElement;
+
+  constructor(private elementRef: ElementRef) { }
+
+  ngAfterContentInit() {
+    this.announcerEl = DOM.querySelector(this.elementRef.nativeElement, '._md-aria-announcer');
+  }
+
+  announce(message: string): void {
+    this.announcerEl.textContent = '';
```
Comment (jelbourn) [inline src/components/aria-live/aria-live.ts]: Did it not work to simply put a `{{ }}` binding inside the `aria-live` element, where the bound property is only updated after a timeout?

File: src/components/aria-live/aria-live.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+    // The most screen-readers need a delay after the clear action, to probably recognise the
+    // announcement. 100ms is a average delay, which works for the most supported readers.
```
Comment (jelbourn) [inline src/components/aria-live/aria-live.ts]: Can you use my comment from my original code verbatim?
```
  // This 100ms timeout is necessary for some browser + screen-reader combinations:
  // - Both JAWS and NVDA over IE11 will not announce anything without a non-zero timeout.
  // - With Chrome and IE11 with NVDA or JAWS, a repeated (identical) message won't be read a
  //   second time without clearing and then using a non-zero delay.
  // (using JAWS 17 at time of this writing).
```

File: src/components/aria-live/aria-live.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+  private announcerEl: HTMLElement;
+
+  constructor(private elementRef: ElementRef) { }
+
+  ngAfterContentInit() {
+    this.announcerEl = DOM.querySelector(this.elementRef.nativeElement, '._md-aria-announcer');
+  }
+
+  announce(message: string): void {
```
Comment (jelbourn) [inline src/components/aria-live/aria-live.ts]: It would be good if the announce function took a second, optional argument for the politeness.

File: src/components/aria-live/aria-live.spec.ts [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+  describe('MdAriaLive', () => {
+    let builder: TestComponentBuilder;
+    let timeoutMock: any;
+
+    beforeEach(inject([TestComponentBuilder], (tcb: TestComponentBuilder) => {
+      builder = tcb;
+    }));
+
+    beforeAll(() => {
+      timeoutMock = mockTimeout();
```
Comment (jelbourn) [inline src/components/aria-live/aria-live.spec.ts]: You should be able to do this more easily with Angular 2's `fakeAsync`; check out the portal tests for an example.

File: src/components/aria-live/README.md [hunk 8]
```diff
@@ -0,0 +1,37 @@
+# MdAriaLive
```
Comment (jelbourn) [inline src/components/aria-live/README.md]: I kind of want to call this `md-live-announcer`. 

@hansl @kara @robertmesserle thoughts?

File: src/components/aria-live/aria-live.spec.ts [hunk 9]
```diff
@@ -0,0 +1,95 @@
+import {
```
Comment (jelbourn) [inline src/components/aria-live/aria-live.spec.ts]: This may seem weird, but could you move this component into `core/` for now? I don't yet want to publish it as a first-class component.

**Comment by jelbourn** [comment]
cc @marcysutton in case you're interested :)

File: src/components/aria-live/README.md (hunk 8 above)
Comment (hansl) [inline src/components/aria-live/README.md]: Yes, MdLiveAnnouncer would probably be better. Do we want this to be a component or a service (or both)?

Comment (jelbourn) [inline src/components/aria-live/README.md]: I believe this does make the most sense as a component, since you can inject it both ways.

File: src/components/aria-live/aria-live.spec.ts (hunk 9 above)
Comment (DevVersion) [inline src/components/aria-live/aria-live.spec.ts]: Can do that, but it seems like you're only compiling styles within `src/core/styles`, I can add the component to BrocolliSass as well, or we keep it temporary in the `src/components`

File: src/components/aria-live/aria-live.scss (hunk 1 above)
Comment (DevVersion) [inline src/components/aria-live/aria-live.scss]: Yes right, seems to be unnecessary, because of the view encapsulation.

File: src/components/aria-live/aria-live.scss (hunk 2 above)
Comment (DevVersion) [inline src/components/aria-live/aria-live.scss]: Done.

File: src/components/aria-live/aria-live.html (hunk 3 above)
Comment (DevVersion) [inline src/components/aria-live/aria-live.html]: Oh, if it has no `ng-content`, then it can't be a parent component. That would mean, we are not able to inject the component. 

Otherwise we need to provide a global OpaqueToken or something else. 
In my opinion, injecting it as the parent component is very elegant.

What are your thoughts?

File: src/components/aria-live/aria-live.ts (hunk 4 above)
Comment (DevVersion) [inline src/components/aria-live/aria-live.ts]: I tested that out, but when we we use a bound text property, then we have to set the property always to undefined, to clear the text content, setting to `''` isn't working for me in this case. (Voiceover iOS)

Additionally, I think, that a change detection is here not necessary. 
If you have other thoughts let me know, and I'll change it to Angular's Property Bound.

File: src/components/aria-live/aria-live.spec.ts (hunk 7 above)
Comment (DevVersion) [inline src/components/aria-live/aria-live.spec.ts]: Awesome, that's what I've been searching for.

File: src/components/aria-live/README.md (hunk 8 above)
Comment (DevVersion) [inline src/components/aria-live/README.md]: Changed it to `MdLiveAnnouncer`.

File: src/components/aria-live/aria-live.ts (hunk 5 above)
Comment (DevVersion) [inline src/components/aria-live/aria-live.ts]: Done

File: src/components/aria-live/aria-live.ts (hunk 6 above)
Comment (DevVersion) [inline src/components/aria-live/aria-live.ts]: Good idea, will implement that.

File: src/components/live-announcer/README.md [hunk 10]
```diff
@@ -0,0 +1,37 @@
+# MdAriaLive
```
Comment (jelbourn) [inline src/components/live-announcer/README.md]: The readme still uses `aria-live`

File: src/components/live-announcer/live-announcer.ts [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+  styleUrls: ['./components/live-announcer/live-announcer.css'],
+})
+export class MdLiveAnnouncer implements AfterContentInit {
+
+  private announcerEl: HTMLElement;
+
+  constructor(private elementRef: ElementRef) { }
+
+  ngAfterContentInit() {
+    this.announcerEl = DOM.querySelector(this.elementRef.nativeElement, '.md-live-announcer');
```
Comment (jelbourn) [inline src/components/live-announcer/live-announcer.ts]: Why did you change this away from using the `Renderer` ? If the binding doesn't work, using the `Renderer` is preferable.

File: src/components/live-announcer/README.md (hunk 10 above)
Comment (DevVersion) [inline src/components/live-announcer/README.md]: Oh, shameful. I was sure to change that.

**Comment by jelbourn** [comment]
@DevVersion sorry to change course on you with this PR, but I talked with hansl@ and we realized that doing this as a service would actually make more sense. 
* It ensures that there's only one announcer globally
* It can be injected anywhere in the app without having to be a parent of anything

The service would be responsible for making its own DOM element and appending it to the document body (and also checking whether there is already an announcer element present).

**Comment by DevVersion** [comment]
@jelbourn That's no problem. I will change that as soon as possible and ping you. (lot of stress atm)

**Comment by jelbourn** [comment]
@DevVersion No worries! We don't have any immediate need for this.

**Comment by DevVersion** [comment]
@jelbourn Updated it. Please take a look again.

File: src/core/live-announcer/live-announcer.ts [hunk 12]
```diff
@@ -0,0 +1,52 @@
+import {Injectable} from 'angular2/core';
+import {DOM} from '../../core/platform/dom/dom_adapter';
+
+@Injectable()
+export class MdLiveAnnouncer {
+
+  private announcerEl: Element;
```
Comment (jelbourn) [inline src/core/live-announcer/live-announcer.ts]: How about `_liveElement` ?

File: src/core/live-announcer/live-announcer.ts [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    // This 100ms timeout is necessary for some browser + screen-reader combinations:
+    // - Both JAWS and NVDA over IE11 will not announce anything without a non-zero timeout.
+    // - With Chrome and IE11 with NVDA or JAWS, a repeated (identical) message won't be read a
+    //   second time without clearing and then using a non-zero delay.
+    // (using JAWS 17 at time of this writing).
+    setTimeout(() => this.announcerEl.textContent = message, 100);
+  }
+
+  private _getAnnouncerElement(): Element {
+    let documentBody = DOM.getGlobalEventTarget('body');
```
Comment (jelbourn) [inline src/core/live-announcer/live-announcer.ts]: Since we decided to drop TS support yesterday, you can simply use `document.createElement`, etc.

Instead of looking for an existing announcer element, the service should just always create one.

File: src/core/live-announcer/live-announcer.ts [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+
+  constructor() {
+    this.announcerEl = this._getAnnouncerElement();
+  }
+
+  /**
+   * @param message Message to be announced to the screenreader
+   * @param politeness The politeness of the announcer element.
+   */
+  announce(message: string, politeness = 'polite'): void {
```
Comment (jelbourn) [inline src/core/live-announcer/live-announcer.ts]: Cool TypeScript trick: you can introduce your own type for the politeness

```
// higher up
export type AriaLivePoliteness: 'off' | 'polite' | 'assertive';

// then, here
 announce(message: string, politeness: AriaLivePoliteness = 'polite'): void
```
This enforces that people only provide a valid politeness

File: src/core/live-announcer/live-announcer.ts (hunk 13 above)
Comment (marcysutton) [inline src/core/live-announcer/live-announcer.ts]: Just make sure any straggling elements get cleaned up so there aren't conflicts.

File: src/core/live-announcer/live-announcer.ts [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+    this.announcerEl.textContent = '';
+
+    this.announcerEl.setAttribute('aria-live', politeness);
```
Comment (marcysutton) [inline src/core/live-announcer/live-announcer.ts]: I think we'll need to validate that a changing politeness level is correctly utilized by Assistive Technologies.

File: src/core/live-announcer/live-announcer.ts (hunk 13 above)
Comment (jelbourn) [inline src/core/live-announcer/live-announcer.ts]: The service is meant to be a singleton, so it should only ever create one element for itself. It might make sense to add another API later to pass it an existing `aria-live` element to use instead of creating its own.

Comment (DevVersion) [inline src/core/live-announcer/live-announcer.ts]: Yes we can do this later, by providing an OpaqueToken IMO.

File: src/core/live-announcer/live-announcer.ts (hunk 15 above)
Comment (jelbourn) [inline src/core/live-announcer/live-announcer.ts]: That's a good point; @DevVersion can you add a `TODO`:
```ts
TODO: ensure changing the politness works on all environments we support.
```

**Comment by jelbourn** [comment]
LGTM

**Comment by jelbourn** [comment]
@DevVersion Just notices one last thing from CI: IE11 doesn't support `Element.prototype.remove`. Once that's resolved I can merge
